ast/desugar/Desugar.cc

```diff
@@ -508,7 +510,51 @@ TreePtr node2TreeImpl(DesugarContext dctx, unique_ptr<parser::Node> what) {
                     rec = MK::Self(loc.copyWithZeroLength());
                     flags.isPrivateOk = true;
                 }
+
                 if (absl::c_any_of(send->args, [](auto &arg) { return parser::isa_node<parser::Splat>(arg.get()); })) {
+                    // Build up an array that represents the keyword args for the send. When there is a Kwsplat, treat
+                    // all keyword arguments as a single argument.
+                    unique_ptr<parser::Node> kwArray;
+                    if (!send->args.empty()) {
+                        // Deconstruct the kwargs hash in the last argument if it's present.
+                        if (auto *hash = parser::cast_node<parser::Hash>(send->args.back().get())) {
+                            if (hash->kwargs) {
+                                // hold a reference to the node, and remove it from the back fo the send list
+                                auto node = std::move(send->args.back());
+                                send->args.pop_back();
+
+                                parser::NodeVec elts;
+
+                                // skip inlining the kwargs if there are any non-key/value pairs present
+                                if (absl::c_any_of(hash->pairs, [](auto &arg) {
+                                        return !parser::isa_node<parser::Pair>(arg.get());
+                                    })) {
+                                    elts.emplace_back(std::move(node));
+                                } else {
+                                    // inline the hash into the send args
+                                    for (auto &entry : hash->pairs) {
+                                        typecase(
+                                            entry.get(),
+                                            [&](parser::Pair *pair) {
+                                                elts.emplace_back(std::move(pair->key));
+                                                elts.emplace_back(std::move(pair->value));
+                                            },
+                                            [&](parser::Node *node) { Exception::raise("Unhandled case"); });
+                                    }
+                                }
+
+                                kwArray = make_unique<parser::Array>(loc, std::move(elts));
+                            }
+                        }
+                    }
+
+                    // If the kwargs hash is not present, make a `nil` to put in the place of that argument. This
+                    // will be used in the implementation of the intrinsic to tell the difference between keyword
+                    // args, keyword args with kw splats, and no keyword args at all.
+                    if (kwArray == nullptr) {
+                        kwArray = make_unique<parser::Nil>(loc);
+                    }
```

Comment:
It seems unfortunate how similar the two cases for keyword args are in the `Send` case, but they're just different enough that it seemed easier to inline the handling here.